### PR TITLE
chore: add internal benchmarking script

### PR DESCRIPTION
**Pull request by cojenco**
- chore: adds internal benchmarking script (not officially supported)
- chore: move benchwrapper to subdirectory (not officially supported)

### CODE REVIEW

**Review by frankyn — CHANGES_REQUESTED** [review]

File: tests/perf/benchmarking.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    update_wrapper(partial_func, func)
+    return partial_func
+
+
+def generate_func_list(bucket_name, min_size, max_size):
+    """Generate Write-1-Read-3 workload."""
+    blob_name = uuid.uuid4().hex
+    # generate randmon size payload using a uniform dist
+    size = random.randrange(min_size, max_size)
+    payload = _create_block(size)
```
[the remark this answers is not in the document]
Comment (cojenco) [reply tests/perf/benchmarking.py]: Updated to upload from disk using NamedTemporaryFile

File: tests/perf/benchmarking.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+        res["ElapsedTimeUs"],
+        res["CpuTimeUs"],
+        res["Status"],
+        res["RunID"],
+    ]
+
+
+def WRITE(bucket, blob_name, payload, **kwargs):
+    """Perform an upload."""
+    checksum = kwargs.get("checksum")
```
[the remark this answers is not in the document]
Comment (cojenco) [reply tests/perf/benchmarking.py]: yep missed that, thanks!

File: tests/perf/benchmarking.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        res["Status"] = ["FAIL"]
+    else:
+        res["Status"] = ["OK"]
+    end_time = time.monotonic_ns()
+
+    res["ElapsedTimeUs"] = round(
+        (end_time - start_time) / 1000
+    )  # convert nanoseconds to microseconds
+    res["ObjectSize"] = func.keywords.get("size")
+    res["Crc32cEnabled"] = func.keywords.get("Crc32cEnabled")
```
[the remark this answers is not in the document]
Comment (cojenco) [reply tests/perf/benchmarking.py]: makes sense, updated!

File: tests/perf/benchmarking.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cojenco) [reply tests/perf/benchmarking.py]: Changed to download to file as well

File: tests/perf/benchmarking.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+def generate_func_list(bucket_name, min_size, max_size):
+    """Generate Write-1-Read-3 workload."""
+    blob_name = uuid.uuid4().hex
+    # generate randmon size payload using a uniform dist
+    size = random.randrange(min_size, max_size)
+    payload = _create_block(size)
+
+    # generate random checksumming type using a uniform dist
+    idx_checksum = random.randrange(0, 2)
```
[the remark this answers is not in the document]
Comment (cojenco) [reply tests/perf/benchmarking.py]: done

File: tests/perf/benchmarking.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+def measure_performance(func):
+    """Measure latency and throughput per operation call."""
+    # Holds benchmarking results for each operation
+    res = {
+        "ApiName": "JSON",
+        "RunID": TIMESTAMP,
+        "CpuTimeUs": -1,
+        "AppBufferSize": 1024,
+        "LibBufferSize": 104857600,
```
[the remark this answers is not in the document]
Comment (cojenco) [reply tests/perf/benchmarking.py]: `LibBufferSize` is the library default of 100 MiB. Added a constant `DEFAULT_LIB_BUFFER_SIZE` along with a link to docs.

`AppBufferSize` is not supported here, updated!

File: tests/perf/benchmarking.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+        bucket.delete(force=True)
+    except Exception as e:
+        logging.exception(f"Caught an exception while running retry instructions\n {e}")
+
+    return results
+
+
+if __name__ == "__main__":
+    p = multiprocessing.Pool(NUM_PROCESSES)
+    pool_output = p.map(main_profiling, range(1))
```
[the remark this answers is not in the document]
Comment (cojenco) [reply tests/perf/benchmarking.py]: Good catch. I've updated to set number of samples as the iterable in `map` so we're running on multiple processes concurrently. Let me know what other ways can be done.

**Review by frankyn — CHANGES_REQUESTED** [review]

File: tests/perf/benchmarking.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    help="File to output results to",
+)
+args = parser.parse_args()
+
+NUM_SAMPLES = args.num_samples
+NUM_PROCESSES = args.p
+MIN_SIZE = args.min_size
+MAX_SIZE = args.max_size
+BUCKET_LOCATION = args.r
+CSV_PATH = args.o
```
[the remark this answers is not in the document]
Comment (cojenco) [reply tests/perf/benchmarking.py]: Done, also added a `main()` to run the `benchmarker_runner` and output csv file

File: tests/perf/README.md [hunk 8]
```diff
@@ -1,21 +1,45 @@
...
+This runs 10K iterations of Write-1-Read-3 on 5KiB to 16KiB files, and generates output to a default csv file `benchmarking<TIMESTAMP>.csv`:
 ```bash
-$ export STORAGE_EMULATOR_HOST=http://localhost:8080
-$ pip install grpcio
-$ cd storage
+$ cd python-storage
 $ pip install -e . # install google.cloud.storage locally
 $ cd tests/perf
-$ python3 benchwrapper.py --port 8081
+$ python3 benchmarking.py --num_samples 10000 --max_size 16384
```
Comment (Breathtender) [inline tests/perf/README.md:14]: Oooh love the configurability you got in here!

**Review by unforced — APPROVED** [review]
Wow Cathy, this has become really beautifully polished and put together! Thanks so much for continuing to evolve on our initial work together and bringing it to such a well done finished piece!

LGTM

**Review by frankyn — APPROVED** [review]

**Pull request merged**